### PR TITLE
[WIP] AboutLibraries v8 (Gradle Plugin)

### PR DESCRIPTION
**Pull request by mikepenz**
# Changes
* update README to reflect the new seperate module for definition files 
* update README to highlight the new gradle plugin
* write new gradle plugin (and all the relevant code to publish it to bintray)
  * gradle plugin will detect all libraries used as dependency in the project
  * add license information to the AboutLibraries UI for those used
* first fully working state of the gradle plugin approach

# Note
I would love to enhance the information from the POM file with information and details from the already existing definition files

### Ideas
* map gradle ID to library via different file
* introduce new field in definitions to define gradle id

### CODE REVIEW

**Comment by mikepenz** [comment]
This new release also fixes: https://github.com/mikepenz/AboutLibraries/issues/438

**Comment by mikepenz** [comment]
# How to use

## Gradle Plugin

As a new feature of the AboutLibraries v8.x.y we offer a gradle plugin which will resolve the dependency during compilation, and only includes the libraries which are really specified as dependencies.

```javascript
// Root build.gradle
maven {
    url uri('https://dl.bintray.com/mikepenz/gradle-plugins/') // TODO wait for approval in jCenter
}
classpath "com.mikepenz.aboutlibraries.plugin:aboutlibraries-plugin:0.1.1"

// App build.gradle
apply plugin: 'aboutlibraries-plugin'
```

## Using Maven
The AboutLibraries Library is pushed to [Maven Central](http://search.maven.org/#search|ga|1|g%3A%22com.mikepenz%22).

## CORE module

```javascript
implementation "com.mikepenz:aboutlibraries-core:${latestAboutLibsRelease}"
```

## UI module

```javascript
implementation "com.mikepenz:aboutlibraries:${latestAboutLibsRelease}"

//required support lib modules
implementation "androidx.appcompat:appcompat:${versions.appcompat | 1.x.y}"
implementation "androidx.cardview:cardview:${versions.cardview | 1.x.y}"
implementation "androidx.recyclerview:recyclerview:${versions.recyclerview | 1.1.y}"
implementation "com.google.android.material:material:${versions.material | 1.1.y}"
```

**Comment by mikepenz** [comment]
TODO:

- enhance lib from gradle plugin with info from definition file
- task to output all dependencies with their aboutLibs ID
- clean task
- combine libraries like aboutlibraries where we got a core dependency (do not necessarily show it twice?)

**Comment by Kurdenes21** [comment]
# ## ### ** _ > ` [ - @mikepenz  ](url) ` _ **

File: plugin/src/main/groovy/com/mikepenz/aboutlibraries/plugin/AboutLibrariesTask.groovy [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+        if (checkEmpty(authorWebsite)) {
+            outputFile.append("<string name=\"library_${uniqueId}_authorWebsite\">${authorWebsite}</string>")
+        }
+        outputFile.append("<string name=\"library_${uniqueId}_libraryName\">${libraryName}</string>")
+        outputFile.append("<string name=\"library_${uniqueId}_libraryDescription\"><![CDATA[${libraryDescription}]]></string>")
+        outputFile.append("<string name=\"library_${uniqueId}_libraryVersion\">${libraryVersion}</string>")
+        outputFile.append("<string name=\"library_${uniqueId}_libraryArtifactId\">${groupId}:${artifactPom.artifactId}:${artifactPom.version}</string>")
+        // the maven artifact Id
+        if (checkEmpty(libraryWebsite)) {
+            outputFile.append("<string name=\"library_${uniqueId}_libraryWebsiten\">${libraryWebsite}</string>")
```
Comment (koral--) [inline plugin/src/main/groovy/com/mikepenz/aboutlibraries/plugin/AboutLibrariesTask.groovy]:
```suggestion
            outputFile.append("<string name=\"library_${uniqueId}_libraryWebsite\">${libraryWebsite}</string>")
```
Looks like a typo 🤔

File: plugin/src/main/groovy/com/mikepenz/aboutlibraries/plugin/AboutLibrariesTask.groovy [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        }
+        if (checkEmpty(licenseId)) {
+            outputFile.append("<string name=\"library_${uniqueId}_licenseId\">${licenseId}</string>")
```
Comment (koral--) [inline plugin/src/main/groovy/com/mikepenz/aboutlibraries/plugin/AboutLibrariesTask.groovy]: I would rather use some XML building API, Groovy has it built-in IIRC.

File: library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt [hunk 3]
```diff
@@ -11,8 +11,13 @@ import com.mikepenz.aboutlibraries.util.*
 import java.util.*
 import kotlin.collections.ArrayList
 
-public class Libs(context: Context, fields: Array<String> = context.getFields()) {
+public class Libs(
+        context: Context,
+        fields: Array<String> = context.getFields(),
+        libraryEnchantments: HashMap<String, String> = HashMap()
```
Comment (AllanWang) [inline library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt]: Take in `Map`, unless `HashMap` is required.

Comment (AllanWang) [reply library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt]: Default can be `emptyMap()` or `mutableMapOf()` depending on interface

File: library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt [hunk 4]
```diff
@@ -82,21 +89,44 @@ public class Libs(context: Context, fields: Array<String> = context.getFields())
...
-                library.isInternal = true
-                internLibraries.add(library)
-            }
-        }
 
-        //add external libs
-        for (externalIdentifier in foundExternalLibraryIdentifiers) {
-            val library = genLibrary(context, externalIdentifier)
+        //add plugin libs
+        for (pluginLibraryIdentifier in foundPluginLibraryIdentifiers) {
```
Comment (AllanWang) [inline library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt]: This is a matter of personal preference, but there are several ways to write this in a potentially more kotlinized way.

The use of null checks with if statements, or even `T?.let { t -> }` results in a lot of nested braces.
For this one, you can consider adding a separate function (inner function if required) of the form:

```kt
fun someFunctionName(libIdentifier: Type): Library? {
   val library = genLibrary(context, libIdentifier) ?: return null
   ...
   val enchantWithKey = ... ?: return library
   val enchantWith = ... ?: return library
   library.enchantBy(enchantWith)
   return library
}
```

and then do something like

```kt
externLibraries = foundPluginLibraryIdentifiers.mapNotNull { someFunctionName(it) }
```

(or `addAll` if you require mutation)

File: library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt [hunk 5]
```diff
@@ -426,6 +456,9 @@ public class Libs(context: Context, fields: Array<String> = context.getFields())
         if (customVariablesString.isBlank()) {
             customVariablesString = ctx.getStringResourceByName(DEFINE_INT + libraryName)
         }
+        if (customVariablesString.isBlank()) {
```
Comment (AllanWang) [inline library-core/src/main/java/com/mikepenz/aboutlibraries/Libs.kt]: You don't need mutations if you wish.

```kt
val customVariablesString = 
  sequenceOf(DEFINE_EXT, DEFINE_INT, DEFINE_PLUGIN)
    .map { ctx.getResourceByName("$it$libraryName") }
    .filter { it.isNotBlank() /* you used both blank and empty. pick the right one here */ }
    .firstOrNull()
    ?: "" // fallback if necessary
```

The other more general example is

```kt
val a = funA().takeIf { conditionA(it) } ?:  funB().takeIf { conditionB(it) } ?: ...
```

File: library-core/src/main/java/com/mikepenz/aboutlibraries/entity/Library.kt [hunk 6]
```diff
@@ -4,6 +4,7 @@ package com.mikepenz.aboutlibraries.entity
 data class Library(
         var definedName: String,
         var isInternal: Boolean = false,
+        var isPlugin: Boolean = false,
```
Comment (AllanWang) [inline library-core/src/main/java/com/mikepenz/aboutlibraries/entity/Library.kt]: In general, prefer immutable data types where possible. Immutable changes can be done via `copy(paramName = paramValue)`

File: library/src/main/java/com/mikepenz/aboutlibraries/LibsBuilder.kt [hunk 7]
```diff
@@ -374,6 +375,18 @@ class LibsBuilder : Serializable {
         return this
```
Comment (AllanWang) [inline library/src/main/java/com/mikepenz/aboutlibraries/LibsBuilder.kt]: Up to you, but in general you can make these variables public and allow users to modify it via `.apply`. Builders may be redundant in that situation

**Comment by mikepenz** [comment]
@AllanWang thanks for the input :)

File: library/src/main/java/com/mikepenz/aboutlibraries/LibsBuilder.kt (hunk 7 above)
Comment (mikepenz) [reply library/src/main/java/com/mikepenz/aboutlibraries/LibsBuilder.kt]: this is very much still legacy code :/

File: README.md [hunk 8]
```diff
@@ -66,16 +79,20 @@ implementation "androidx.recyclerview:recyclerview:${versions.recyclerview | 1.1
...
-To use java version please use a version smaller than 7.0.0 (See the release on GitHub)
-To use appcompat please use a version smaller than 6.2.0. (See the releases on GitHub)
+## WITHOUT gradle plugin (not recommended)
 
-Further information and how to use it if you can't update to the newest support libs can be found in the [wiki](https://github.com/mikepenz/AboutLibraries/wiki/HOWTO:-Include)
+If you do not want to use the gradle plugin, you need to add the legacy definition files, which will then be included in the built apk, and resolved via reflection during runtime.
+> NOTE: This is not recommended. Please migrate to use the gradle plugin
 
-## Usage
+```javascript
```
Comment (koral--) [inline README.md]:
```suggestion
```gradle
```

File: library-definitions/proguard-rules.txt [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# Add project specific ProGuard rules here.
```
Comment (koral--) [inline library-definitions/proguard-rules.txt]: I would remove this file and reference to it in build.gradle if does not have any used content.

File: plugin/gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.6-all.zip
```
Comment (koral--) [inline plugin/gradle/wrapper/gradle-wrapper.properties]: Is it intended that there is a 2nd wrapper in the project?

Comment (mikepenz) [reply plugin/gradle/wrapper/gradle-wrapper.properties]: yeah the plugin is it's own sub-project

**Pull request merged**